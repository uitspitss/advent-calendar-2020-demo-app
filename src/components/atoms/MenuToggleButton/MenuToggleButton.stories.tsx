import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import base from 'paths.macro';

import { MenuToggleButton, MenuToggleButtonProps } from './MenuToggleButton';

const path = base.replace(
  /^.+\/(atoms|molecules|organisms|templates|pages)\/(.+)\//,
  '$1/$2',
);

export default {
  title: `${path}`,
  component: MenuToggleButton,
  argTypes: {
    isOpen: { control: 'boolean' },
    handleClick: { action: 'clicked' },
  },
} as Meta;

const Template: Story<MenuToggleButtonProps> = (args) => (
  <MenuToggleButton {...args} />
);

export const Closed = Template.bind({});
Closed.args = {
  isOpen: false,
};

export const Open = Template.bind({});
Open.args = {
  isOpen: true,
};
