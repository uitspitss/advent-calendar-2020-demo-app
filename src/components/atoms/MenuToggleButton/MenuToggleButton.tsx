import React, { FC, ComponentPropsWithRef } from 'react';
import tw, { styled } from 'twin.macro';

export type MenuToggleButtonProps = {
  children?: never;
  className?: string;
  /**
   * トグルステータス
   */
  isOpen: boolean;
  /**
   * クリックハンドラ
   */
  handleClick: () => void;
} & ComponentPropsWithRef<'button'>;

type Props = MenuToggleButtonProps;

const Component: FC<Props> = (props) => {
  const { className, handleClick } = props;

  return (
    <button className={className} type="button" onClick={handleClick}>
      <svg
        xmlns="http://www.w3.org/2000/svg"
        fill="none"
        viewBox="0 0 24 24"
        stroke="currentColor"
      >
        <path
          className="open"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          d="M6 18L18 6M6 6l12 12"
        />
        <path
          className="closed"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          d="M4 6h16M4 12h16M4 18h16"
        />
      </svg>
    </button>
  );
};

const StyledComponent = styled(Component)<Pick<Props, 'isOpen'>>`
  ${tw`flex items-center px-3 py-2 border rounded bg-transparent border-0`}

  svg {
    ${tw`w-6 h-6`}
    path {
      ${tw`transition duration-100 ease-in`}
      &.open {
        ${({ isOpen }) => (isOpen ? tw`opacity-100` : tw`opacity-0`)}
      }
      &.closed {
        ${({ isOpen }) => (isOpen ? tw`opacity-0` : tw`opacity-100`)}
      }
    }
  }
`;

/**
 * ヘッダーのモバイルレイアウトでのトグルボタン
 */
export const MenuToggleButton: FC<MenuToggleButtonProps> = (props) => (
  <StyledComponent {...props} />
);
