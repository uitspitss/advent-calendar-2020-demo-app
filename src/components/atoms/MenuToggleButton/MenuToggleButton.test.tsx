import React from 'react';
import renderer from 'react-test-renderer';

import { MenuToggleButton } from './MenuToggleButton';

describe('MenuToggleButton', () => {
  test('Snap Shot', () => {
    const isOpen = false;
    const handleClick = () => {};
    const tree = renderer
      .create(<MenuToggleButton isOpen={isOpen} handleClick={handleClick} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
