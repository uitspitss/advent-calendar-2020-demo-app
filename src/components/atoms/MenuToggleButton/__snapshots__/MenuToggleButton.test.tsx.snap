// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuToggleButton Snap Shot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  border-width: 0;
  border-radius: 0.25rem;
  background-color: transparent;
}

.emotion-0 svg {
  width: 1.5rem;
  height: 1.5rem;
}

.emotion-0 svg path {
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,-webkit-transform;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  -webkit-transition-duration: 100ms;
  transition-duration: 100ms;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,1,1);
  transition-timing-function: cubic-bezier(0.4,0,1,1);
}

.emotion-0 svg path.open {
  opacity: 0;
}

.emotion-0 svg path.closed {
  opacity: 1;
}

<button
  className="emotion-0 emotion-1"
  onClick={[Function]}
  type="button"
>
  <svg
    fill="none"
    stroke="currentColor"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      className="open"
      d="M6 18L18 6M6 6l12 12"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
    />
    <path
      className="closed"
      d="M4 6h16M4 12h16M4 18h16"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
    />
  </svg>
</button>
`;
